[Navigation] Implement the initialize the navigation API entries algorithm

rwlbuis opened the pull request:
#### 4d20880a9333b70b8444b5305d52035589e893ba
<pre>
[Navigation] Implement the initialize the navigation API entries algorithm
<a href="https://bugs.webkit.org/show_bug.cgi?id=267636">https://bugs.webkit.org/show_bug.cgi?id=267636</a>

Reviewed by Alex Christensen and Chris Dumez.

Implement the initialize the navigation API entries algorithm [1] by calling it in FrameLoader
when a new document is starting to be loaded.

Also add a minimal implementation of the has entries and events disabled algorithm [2]
in order to not regress existing navigation API tests.

[1] <a href="https://html.spec.whatwg.org/multipage/nav-history-apis.html#initialize-the-navigation-api-entries-for-a-new-document">https://html.spec.whatwg.org/multipage/nav-history-apis.html#initialize-the-navigation-api-entries-for-a-new-document</a>
[2] <a href="https://html.spec.whatwg.org/multipage/nav-history-apis.html#has-entries-and-events-disabled">https://html.spec.whatwg.org/multipage/nav-history-apis.html#has-entries-and-events-disabled</a>

* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::didBeginDocument):
(WebCore::FrameLoader::updateNavigationAPIEntries):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::navigation):
(WebCore::LocalDOMWindow::protectedNavigation):
* Source/WebCore/page/LocalDOMWindow.h:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::Navigation):
(WebCore::Navigation::initializeEntries):
(WebCore::Navigation::entries const):
(WebCore::Navigation::currentEntry const):
(WebCore::Navigation::navigate):
(WebCore::Navigation::reload):
(WebCore::Navigation::traverseTo):
(WebCore::Navigation::back):
(WebCore::Navigation::forward):
(WebCore::Navigation::hasEntriesAndEventsDisabled const):
* Source/WebCore/page/Navigation.h:
* Source/WebCore/page/NavigationHistoryEntry.cpp:
(WebCore::NavigationHistoryEntry::NavigationHistoryEntry):
* Source/WebCore/page/NavigationHistoryEntry.h:

Canonical link: <a href="https://commits.webkit.org/273532@main">https://commits.webkit.org/273532@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/155c4a1cb8d9bac4bf8b59b1a407d53d2101a5e6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35653 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/14593 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37792 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38392 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32116 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/36852 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/17017 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11619 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30902 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36206 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12336 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31729 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10841 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10843 "Unexpected infrastructure issue, retrying build") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31869 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39637 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32399 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32205 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36801 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11035 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8927 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34882 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12753 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8146 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11543 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11821 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

TingPing commented on Source/WebCore/page/NavigationHistoryEntry.h:
```diff
@@ -42,26 +43,26 @@ class NavigationHistoryEntry final : public RefCounted<NavigationHistoryEntry>,
-    const String& key() const { return m_key; };
-    const String& id() const { return m_id; };
+    const String& url() const { return m_url.string(); };
+    String key() const { return m_key.toString(); };
```
In my later patch this is called in a loop to lookup entries. Maybe the string could be cached, or the primary way it's stored?

rwlbuis replied:
Yeah, this is a first stab at the key/id handling, it can definitely change later, as you said for example for efficiency. I mainly added it for now for identification/debug reasons (so far that means through WebKit inspector, probably later for WPT tests).

achristensen07 commented on Source/WebCore/page/Navigation.h:
```diff
@@ -100,7 +102,9 @@ class Navigation final : public RefCounted<Navigation>, public EventTarget, publ
-    RefPtr<NavigationHistoryEntry> m_currentEntry;
+    bool hasEntriesAndEventsDisabled() const;
+
+    int m_currentEntryIndex { -1 };
```
Is -1 a valid value?  Maybe use std::optional here instead.

rwlbuis replied:
It is kind of implied by the spec:
https://html.spec.whatwg.org/multipage/nav-history-apis.html#getting-the-navigation-api-entry-index

I'll keep it in mind though if it turns out to be more convenient as we do more Navigation API work.

achristensen07 commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -4638,6 +4641,24 @@ RefPtr<DocumentLoader> FrameLoader::loaderForWebsitePolicies(CanIncludeCurrentDo
+{
+    if (!m_frame->document() || !m_frame->document()->settings().navigationAPIEnabled())
+        return;
+    if (auto domWindow = m_frame->document()->domWindow()) {
```
More early returns would make less deeply nested code.

cdumez replied:
Also please store the domWindow in a RefPtr.

rwlbuis replied:
Done.

rwlbuis replied:
And indeed it is much easier on the eyes.

achristensen07 commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -4638,6 +4641,24 @@ RefPtr<DocumentLoader> FrameLoader::loaderForWebsitePolicies(CanIncludeCurrentDo
+            if (RefPtr currentItem = page->backForward().currentItem()) {
+                Vector<Ref<HistoryItem>> historyItems;
+                for (int index = 0; index < (int)page->backForward().count(); index++) {
+                    if (auto item = page->backForward().itemAtIndex(index - page->backForward().backCount()))
```
c-style cast to an int isn't great.  Also subtracting a constant value that needs to be fetched from multiple function calls each time inside a loop isn't great.  I think iterating from -1 * static_cast&lt;int&gt;(backCount()) to and including forwardCount() would be better.

rwlbuis replied:
You are right, I knew the code was quite hacky when I worked on it :} Fixed.

cdumez commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -4638,6 +4641,24 @@ RefPtr<DocumentLoader> FrameLoader::loaderForWebsitePolicies(CanIncludeCurrentDo
+                    if (auto item = page->backForward().itemAtIndex(index - page->backForward().backCount()))
+                        historyItems.append(*item);
+                }
+                domWindow->navigation().initializeEntries(*currentItem, historyItems);
```
please add a `protectedNavigation()` getter to LocalDOMWindow which returns a `Ref<Navigation>` and use this here. We shouldn't be calling a member function on an object we don't hold a strong ref to.

rwlbuis replied:
Done.

cdumez commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -4638,6 +4641,24 @@ RefPtr<DocumentLoader> FrameLoader::loaderForWebsitePolicies(CanIncludeCurrentDo
+            if (RefPtr currentItem = page->backForward().currentItem()) {
+                Vector<Ref<HistoryItem>> historyItems;
+                for (int index = 0; index < (int)page->backForward().count(); index++) {
+                    if (auto item = page->backForward().itemAtIndex(index - page->backForward().backCount()))
```
Please store `item` in a RefPtr.

rwlbuis replied:
Done.

cdumez commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -4638,6 +4641,24 @@ RefPtr<DocumentLoader> FrameLoader::loaderForWebsitePolicies(CanIncludeCurrentDo
+                Vector<Ref<HistoryItem>> historyItems;
+                for (int index = 0; index < (int)page->backForward().count(); index++) {
+                    if (auto item = page->backForward().itemAtIndex(index - page->backForward().backCount()))
+                        historyItems.append(*item);
```
`item.releaseNonNull()`

rwlbuis replied:
Done.

cdumez commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -39,6 +40,28 @@ Navigation::Navigation(ScriptExecutionContext* context, LocalDOMWindow& window)
+void Navigation::initializeEntries(const Ref<HistoryItem>& currentItem, const Vector<Ref<HistoryItem>>& items)
+{
+    for (auto item : items)
+        m_entries.append(NavigationHistoryEntry::create(scriptExecutionContext(), item->url()));
```
`protectedScriptExecutionContext().get()`

rwlbuis replied:
Sure, done.

cdumez commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -39,6 +40,28 @@ Navigation::Navigation(ScriptExecutionContext* context, LocalDOMWindow& window)
+// https://html.spec.whatwg.org/multipage/nav-history-apis.html#initialize-the-navigation-api-entries-for-a-new-document
+void Navigation::initializeEntries(const Ref<HistoryItem>& currentItem, const Vector<Ref<HistoryItem>>& items)
+{
+    for (auto item : items)
```
`auto` -> `Ref`

rwlbuis replied:
Done.

cdumez commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -39,6 +40,28 @@ Navigation::Navigation(ScriptExecutionContext* context, LocalDOMWindow& window)
+    m_currentEntryIndex = items.find(currentItem);
+}
+
+Vector<Ref<NavigationHistoryEntry>> Navigation::entries() const
```
It is not great to be copying a Vector of refpointers whenever this is called. It's a lot of unnecessary churn. If I were you, I'd store an empty vector in a `static NeverDestroyed<Vector<Ref<NavigationHistoryEntry>>>` and return that in the early return case. This way the return type can be `const Vector<Ref<NavigationHistoryEntry>>`

rwlbuis replied:
Fair enough, I added it.

cdumez commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -54,7 +77,7 @@ EventTargetInterface Navigation::eventTargetInterface() const
 {
     // FIXME: keep track of promises to resolve later.
-    auto entry = NavigationHistoryEntry::create(scriptExecutionContext());
+    auto entry = NavigationHistoryEntry::create(scriptExecutionContext(), { });
```
Can you please switch from `auto` to `Ref` since you're updating this line?

rwlbuis replied:
Done.

cdumez commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -65,7 +88,7 @@ Navigation::Result Navigation::navigate(const String& /* url */, NavigateOptions
 {
     // FIXME: keep track of promises to resolve later.
-    auto entry = NavigationHistoryEntry::create(scriptExecutionContext());
+    auto entry = NavigationHistoryEntry::create(scriptExecutionContext(), { });
```
Can you please switch from `auto` to `Ref` since you're updating this line?

rwlbuis replied:
Done.

cdumez commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -76,7 +99,7 @@ Navigation::Result Navigation::reload(ReloadOptions&&, Ref<DeferredPromise>&& co
 {
     // FIXME: keep track of promises to resolve later.
-    auto entry = NavigationHistoryEntry::create(scriptExecutionContext());
+    auto entry = NavigationHistoryEntry::create(scriptExecutionContext(), { });
```
Can you please switch from `auto` to `Ref` since you're updating this line?

rwlbuis replied:
Done.

cdumez commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -87,7 +110,7 @@ Navigation::Result Navigation::traverseTo(const String& /* key */, Options&&, Re
 {
     // FIXME: keep track of promises to resolve later.
-    auto entry = NavigationHistoryEntry::create(scriptExecutionContext());
+    auto entry = NavigationHistoryEntry::create(scriptExecutionContext(), { });
```
Can you please switch from `auto` to `Ref` since you're updating this line?

rwlbuis replied:
Done.

cdumez commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -98,7 +121,7 @@ Navigation::Result Navigation::back(Options&&, Ref<DeferredPromise>&& committed,
 {
     // FIXME: keep track of promises to resolve later.
-    auto entry = NavigationHistoryEntry::create(scriptExecutionContext());
+    auto entry = NavigationHistoryEntry::create(scriptExecutionContext(), { });
```
Can you please switch from `auto` to `Ref` since you're updating this line?

rwlbuis replied:
Done.

cdumez commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -110,5 +133,13 @@ void Navigation::updateCurrentEntry(UpdateCurrentEntryOptions&&)
+// https://html.spec.whatwg.org/multipage/nav-history-apis.html#has-entries-and-events-disabled
+bool Navigation::hasEntriesAndEventsDisabled() const
+{
+    if (window()->securityOrigin() && window()->securityOrigin()->isOpaque())
```
All on one line:
```
return window()->securityOrigin() && window()->securityOrigin()->isOpaque();
```

rwlbuis replied:
Sure, this method will have a lot more if checks so I adopted that pattern for now, but until then I can make it one line.

cdumez commented on Source/WebCore/page/Navigation.h:
```diff
@@ -75,13 +75,15 @@ class Navigation final : public RefCounted<Navigation>, public EventTarget, publ
-    Vector<Ref<NavigationHistoryEntry>> entries() { return m_entries; };
-    RefPtr<NavigationHistoryEntry> currentEntry() { return m_currentEntry; };
+    Vector<Ref<NavigationHistoryEntry>> entries() const;
+    RefPtr<NavigationHistoryEntry> currentEntry() const;
```
This should probably return a `NavigationHistoryEntry*`, there is no ownership transfer here.

rwlbuis replied:
Fixed.

rwlbuis committed:
[Navigation] Implement the initialize the navigation API entries algorithm
https://bugs.webkit.org/show_bug.cgi?id=267636

Reviewed by Alex Christensen and Chris Dumez.

Implement the initialize the navigation API entries algorithm [1] by calling it in FrameLoader
when a new document is starting to be loaded.

Also add a minimal implementation of the has entries and events disabled algorithm [2]
in order to not regress existing navigation API tests.

[1] https://html.spec.whatwg.org/multipage/nav-history-apis.html#initialize-the-navigation-api-entries-for-a-new-document
[2] https://html.spec.whatwg.org/multipage/nav-history-apis.html#has-entries-and-events-disabled

* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::didBeginDocument):
(WebCore::FrameLoader::updateNavigationAPIEntries):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::navigation):
(WebCore::LocalDOMWindow::protectedNavigation):
* Source/WebCore/page/LocalDOMWindow.h:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::Navigation):
(WebCore::Navigation::initializeEntries):
(WebCore::Navigation::entries const):
(WebCore::Navigation::currentEntry const):
(WebCore::Navigation::navigate):
(WebCore::Navigation::reload):
(WebCore::Navigation::traverseTo):
(WebCore::Navigation::back):
(WebCore::Navigation::forward):
(WebCore::Navigation::hasEntriesAndEventsDisabled const):
* Source/WebCore/page/Navigation.h:
* Source/WebCore/page/NavigationHistoryEntry.cpp:
(WebCore::NavigationHistoryEntry::NavigationHistoryEntry):
* Source/WebCore/page/NavigationHistoryEntry.h:

Canonical link: https://commits.webkit.org/273532@main
